drivers/CpuUid.cpp

```diff
@@ -0,0 +1,80 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "hal/cpu_uid_api.h"
+#include "platform/mbed_assert.h"
+#include "drivers/CpuUid.h"
+
+#if DEVICE_CPUUID
+
+namespace mbed {
+
+const char CpuUid::_hexChars[16] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+CpuUid::CpuUid() : _data(NULL)
+{
+    _size = cpu_uid_get_length();
+#ifdef MBED_ASSERT
+    MBED_ASSERT(_size > 0);
+#endif
+    if (0 < _size) {
+        _data = new uint8_t[_size];
+        cpu_uid_get_uid(_data);
+    }
+}
+
+CpuUid::~CpuUid()
+{
+    if (_data) {
+        delete _data;
+    }
+}
+
+CpuUid::operator std::string()
```

Comment:
std::String is not to be part of public APIs. Not sure if we wrote that down here but we should if not. https://os.mbed.com/docs/v5.6/tutorials/arm-mbed-os-software-design-guide.html

Reply:
Why not?
I thought it would be a good idea to provide an interface with std::string, so callers would not have to mind about buffer sizes.

My point is - If in mbed the drivers are all done in C++ - why not use more C++ stuff? ;)

Reply:
We don't need this at all. The size of the stringified output is known at compile time. No C++ stuff is necessary.

I would suggest populating a static buffer with the text representation.

```C
const char* get_c_str() {
    static const char* c_str = NULL;
    static const char buffer[MBED_CPU_UID_STR_SIZE];
    if (c_str == NULL) {
        uint8_t uid_buffer[MBED_CPU_UID_SIZE];
        cpu_uid_get_uid(uid_buffer);
        for (size_t i = 0; i < MBED_CPU_UID_SIZE; ++i) {
            buffer[i*2] = _hexChars[uid_buffer[i] >> 4];
            buffer[i*2+1] = _hexChars[uid_buffer[i] & 0x0F];
        }
        c_str = buffer;
    }
    return c_str;
}
```

Reply:
where should that get_c_str() function belong?
into HAL or in the driver?